### PR TITLE
Add pathfinder

### PR DESCRIPTION
**Pull request by mprevel**
Hi,

This is a pathfinding algorithm as discussed in https://github.com/PurpleKingdomGames/indigo/issues/646

It took me a little bit more time than what I was expecting.

Please let me know if you have questions, comments or want some changes about it.

Regards,

### CODE REVIEW

File: indigo/indigo-extras/src/main/scala/indigoextras/pathfinding/DefaultPathBuilders.scala [hunk 1]
```diff
@@ -0,0 +1,210 @@
+package indigoextras.pathfinding
+
+import indigo.*
+import indigoextras.pathfinding.DefaultPathBuilders.Movements.*
+
+import scala.scalajs.js
+
+// simple path builder implementations based on Point
+object DefaultPathBuilders:
```
Comment (davesmith00000) [inline indigo/indigo-extras/src/main/scala/indigoextras/pathfinding/DefaultPathBuilders.scala:9]: Could this all be in `object PathBuilder`? One less term for users to learn?

File: indigo/indigo-extras/src/main/scala/indigoextras/pathfinding/SearchGrid.scala [hunk 2]
```diff
@@ -6,8 +6,11 @@ import indigoextras.pathfinding.GridSquare.EndSquare
 import indigoextras.pathfinding.GridSquare.ImpassableSquare
 import indigoextras.pathfinding.GridSquare.StartSquare
 
+import scala.annotation.nowarn
 import scala.annotation.tailrec
 
+@deprecated("Use the new indigoextras.pathfinding.PathFinder", "0.15.3")
+@nowarn("cat=deprecation")
 final case class SearchGrid(
```
Comment (davesmith00000) [inline indigo/indigo-extras/src/main/scala/indigoextras/pathfinding/SearchGrid.scala:14]: It's tempting to just remove this stuff. I doubt many people are using it. :thinking:

**Review by davesmith00000 — COMMENTED** [review]
Apologies for the slow review. This looks great! :rocket: 

I've left a few minor comments for your consideration - mostly just for style / consistency. I think the next things are:

1. I'd like to see it working / have an example to turn to for reference. Usually at this point I'd set up a scene in the sandbox, maybe just a simple grid and have it solve and show coloured boxes for the path or something. I can do this if you'd rather not.

2. Decide whether or not to just remove the old search grid. I think we could. Indigo isn't in a place where people are moaning about breaking API changes.

We could just raise issues for those two points, and I can make a demo and then consider removing SearchGrid as steps outside this PR.

File: indigo/indigo-extras/src/main/scala/indigoextras/pathfinding/PathFinder.scala [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  def heuristic(
+      t1: T,
+      t2: T
+  ): Int // heuristic allows to select the way to estimate the distance from a point to the end
+
+// A* (A Star) inspired algorithm version allowing generic types and customisation of the path finding
+object PathFinder:
+
+  private type OpenList[T]   = List[PathProps[T]] // the open list is the list of the points to explore
+  private type ClosedList[T] = List[PathProps[T]] // the closed list is the list of the points already explored
```
Comment (davesmith00000) [inline indigo/indigo-extras/src/main/scala/indigoextras/pathfinding/PathFinder.scala:24]: Avoid `List` where possible, its very slow compared to `Batch`. :smile:

File: indigo/indigo-extras/src/main/scala/indigoextras/pathfinding/PathFinder.scala [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    * @param h
+    *   the heuristic (distance from end)
+    * @param f
+    *   the g + h
+    * @param parent
+    *   the parent point
+    * @tparam T
+    *   the type of the points
+    */
+  private case class PathProps[T](value: T, g: Int, h: Int, f: Int, parent: Option[PathProps[T]] = None)
```
Comment (davesmith00000) [inline indigo/indigo-extras/src/main/scala/indigoextras/pathfinding/PathFinder.scala:40]: Avoid default args, I'd rather be explicit everywhere, almost especially on internal plumbing.

File: indigo/indigo-extras/src/main/scala/indigoextras/pathfinding/SearchGrid.scala (hunk 2 above)
Comment (mprevel) [reply indigo/indigo-extras/src/main/scala/indigoextras/pathfinding/SearchGrid.scala:14]: I did not remove it.

**Comment by mprevel** [comment]
The PR is updated according to the comments but the removal of the old pathfinder.

**Review by davesmith00000 — APPROVED** [review]
Fantastic work @mprevel, and the demo is very cool. :sunglasses: 
image

**Pull request merged**